### DISCUSSION TITLE
Generate commit messages from diff

### DISCUSSION THREAD

**Discussion by dufferzafar** [TheR1D/shell_gpt#293] [Ideas]
There's a bunch of tools around automating the generation of commit messages using GPT.

For eg:

https://github.com/RomanHotsiy/commitgpt/blob/main/template.ts

Being able to do this via shell_gpt itself would be great.

Looking at the code, would I need to create a new role or a new handler?

**Comment by TheR1D** [comment]
To generate git commit message, we can use default Shell GPT role. Just pipe the `git diff` to the `sgpt` and pass simple prompt:
```shell
git diff | sgpt "Generate git commit message for my changes"
```
But if you need something more specific you can make a new role for it:
```
sgpt --create-role git_diff
Enter role description: Suggest 10 commit messages based on the following diff, follow conventional commits, message format should be: <type>[scope]: <description>.
Enter expecting result, e.g. answer, code, shell command, command description, etc.: git commit message
git diff | sgpt --role git_diff

# 1. feat[sgpt]: Add print statements to main function
# 2. chore[sgpt]: Update app.py with new code
# ...
```

**Reply by dufferzafar** [reply]
Ah, didn't know it was that easy! Where is this role information stored on disk?

**Reply by TheR1D** [reply]
`sgpt --list-roles` will show path to "role" files.

**Comment by moritztim** [comment]
```sh
cgpt() {
  commit_msg=$(git diff --staged | sgpt --model="gpt-3.5-turbo" "Write one commit message based on the changes. First letter uppercase, present tense. Your output is directly fed into the command!")
  
  echo -n "Generated Message: \n $commit_msg \n Is this message ok? [y/n]: "
  read confirm
  
  if [[ $confirm == "y" || $confirm == "Y" ]]; then
  git commit -m "$commit_msg"
  fi
}
```
I've just added this to my `~/.zshrc`. I've made it only use the staged files by using `git diff --staged`
